File: C++/reverseList2.cpp
/**
 * @File    : reverseList2.cpp
 * @Brief   : 反转链表
 * @Link    : https://leetcode-cn.com/problems/fan-zhuan-lian-biao-lcof/
 * @Author  : Wei Li
 * @Date    : 2021-09-04
*/

#include <iostream>

// Definition for singly-linked list.
struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution
{
public:
    ListNode *reverseList(ListNode *head)
    {
        ListNode *prev = nullptr;
        ListNode *curr = head;
        while (curr)
        {
            ListNode *next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }
};

// -------------------------------
int main(int argc, char **argv)
{
    // test on leetcode online.

    return 0;
}

File: Python/reverseList2.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: reverseList2.py
@Function: 反转链表
@Link: https://leetcode-cn.com/problems/fan-zhuan-lian-biao-lcof/
@Python Version: 3.8
@Author: Wei Li
@Date：2021-09-04
"""


from typing import List


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        cur, pre = head, None
        while cur:
            tmp = cur.next  # 暂存后继节点 cur.next
            cur.next = pre  # 修改 next 引用指向
            pre = cur      # pre 暂存 cur
            cur = tmp      # cur 访问下一节点
        return pre


# ----------------------------------
if __name__ == "__main__":
    # test on LeetCode online.
    pass
